File: ios/App/Sources/Features/Chats/Content/MathText.swift
import SwiftUI

/// Native latex-lite rendering (iOS twin of MathText.kt) — no WebView, no
/// remote service, no dependencies. Supports \frac, \sqrt, ^{} _{} (and
/// single-token forms), the Greek alphabet and a wide operator set mapped to
/// Unicode. Superscript/subscript ride per-piece baseline offsets in a
/// concatenated Text; Latin letters render italic.
///
/// Graceful degradation is the contract: any UNKNOWN \command makes
/// renderMathText return nil and the caller shows the ORIGINAL expression
/// verbatim — a wrong-looking formula is worse than an honest literal one.

private struct MPiece {
    let text: String
    var script: Int = 0 // 1 superscript, -1 subscript, 0 normal
    var italic: Bool = false
}

private let greekMap: [String: String] = [
    "alpha": "α", "beta": "β", "gamma": "γ", "delta": "δ", "epsilon": "ε",
    "varepsilon": "ε", "zeta": "ζ", "eta": "η", "theta": "θ", "vartheta": "ϑ",
    "iota": "ι", "kappa": "κ", "lambda": "λ", "mu": "μ", "nu": "ν", "xi": "ξ",
    "omicron": "ο", "pi": "π", "varpi": "ϖ", "rho": "ρ", "varrho": "ϱ",
    "sigma": "σ", "varsigma": "ς", "tau": "τ", "upsilon": "υ", "phi": "φ",
    "varphi": "ϕ", "chi": "χ", "psi": "ψ", "omega": "ω",
    "Gamma": "Γ", "Delta": "Δ", "Theta": "Θ", "Lambda": "Λ", "Xi": "Ξ",
    "Pi": "Π", "Sigma": "Σ", "Upsilon": "Υ", "Phi": "Φ", "Psi": "Ψ", "Omega": "Ω"
]

private let operatorMap: [String: String] = [
    "times": "×", "cdot": "·", "div": "÷", "pm": "±", "mp": "∓",
    "leq": "≤", "geq": "≥", "neq": "≠", "ne": "≠", "approx": "≈",
    "equiv": "≡", "sim": "∼", "simeq": "≃", "propto": "∝", "infty": "∞",
    "sum": "∑", "prod": "∏", "int": "∫", "iint": "∬", "oint": "∮",
    "partial": "∂", "nabla": "∇", "in": "∈", "notin": "∉", "subset": "⊂",
    "subseteq": "⊆", "supset": "⊃", "supseteq": "⊇", "cup": "∪", "cap": "∩",
    "forall": "∀", "exists": "∃", "nexists": "∤", "emptyset": "∅",
    "varnothing": "∅", "rightarrow": "→", "to": "→", "leftarrow": "←",
    "gets": "←", "leftrightarrow": "↔", "Rightarrow": "⇒", "Leftarrow": "⇐",
    "Leftrightarrow": "⇔", "mapsto": "↦", "uparrow": "↑", "downarrow": "↓",
    "circ": "∘", "bullet": "•", "star": "⋆", "ast": "∗", "angle": "∠",
    "degree": "°", "perp": "⊥", "parallel": "∥", "land": "∧", "lor": "∨",
    "neg": "¬", "hbar": "ℏ", "ell": "ℓ", "aleph": "ℵ", "wp": "℘",
    "Re": "ℜ", "Im": "ℑ", "prime": "′", "langle": "⟨", "rangle": "⟩",
    "lceil": "⌈", "rceil": "⌉", "lfloor": "⌊", "rfloor": "⌋",
    "cdots": "⋯", "ldots": "…", "vdots": "⋮", "ddots": "⋱", "over": "/"
]

/// Pre-normalization: drop spacing/limits commands and delimiter wrappers.
private func normalizeLatex(_ src: String) -> String {
    src.replacingOccurrences(of: "\\left", with: "")
        .replacingOccurrences(of: "\\right", with: "")
        .replacingOccurrences(of: "\\displaystyle", with: "")
        .replacingOccurrences(of: "\\limits", with: "")
        .replacingOccurrences(of: "\\,", with: " ")
        .replacingOccurrences(of: "\\;", with: " ")
        .replacingOccurrences(of: "\\:", with: " ")
        .replacingOccurrences(of: "\\!", with: "")
}

private struct MathParser {
    let s: String
    var i: String.Index
    var out: [MPiece] = []

    init(_ s: String) {
        self.s = s
        self.i = s.startIndex
    }

    var done: Bool { i >= s.endIndex }

    /// Parses until end (or the matching `}` when stopBrace); false = unsupported.
    mutating func parse(stopBrace: Bool) -> Bool {
        while i < s.endIndex {
            let c = s[i]
            if c == "}" {
                if stopBrace { return true }
                i = s.index(after: i)
            } else if c == "{" {
                i = s.index(after: i)
                if !parse(stopBrace: true) { return false }
            } else if c == "^" || c == "_" {
                i = s.index(after: i)
                let script = c == "^" ? 1 : -1
                var arg: [MPiece] = []
                if !readArg(into: &arg) { return false }
                out.append(contentsOf: arg.map { piece in
                    MPiece(text: piece.text, script: script, italic: piece.italic)
                })
            } else if c == "\\" {
                if !parseCommand() { return false }
            } else if c.isLetter {
                out.append(MPiece(text: String(c), italic: true))
                i = s.index(after: i)
            } else if c.isWhitespace {
                out.append(MPiece(text: " "))
                i = s.index(after: i)
            } else {
                out.append(MPiece(text: String(c)))
                i = s.index(after: i)
            }
        }
        return true
    }

    /// Reads a {group} or single-token argument into sink.
    private mutating func readArg(into sink: inout [MPiece]) -> Bool {
        if i >= s.endIndex { return false }
        if s[i] == "{" {
            i = s.index(after: i)
            let saved = out.count
            if !parse(stopBrace: true) { return false }
            sink.append(contentsOf: out.suffix(from: saved))
            return true
        }
        if s[i] == "\\" {
            let saved = out.count
            if !parseCommand() { return false }
            sink.append(contentsOf: out.suffix(from: saved))
            return true
        }
        sink.append(MPiece(text: String(s[i]), italic: s[i].isLetter))
        i = s.index(after: i)
        return true
    }

    private mutating func parseCommand() -> Bool {
        let after = s.index(after: i)
        if after >= s.endIndex { return false }
        var end = after
        while end < s.endIndex, s[end].isLetter { end = s.index(after: end) }
        let name = String(s[after..<end])
        i = end
        switch name {
        case "frac":
            var a: [MPiece] = []
            var b: [MPiece] = []
            if !readArg(into: &a) || !readArg(into: &b) { return false }
            out.append(MPiece(text: "("))
            out.append(contentsOf: a)
            out.append(MPiece(text: ")/("))
            out.append(contentsOf: b)
            out.append(MPiece(text: ")"))
        case "sqrt":
            // Optional [n] root index → superscript before the radical.
            if i < s.endIndex, s[i] == "[" {
                guard let close = s[i...].firstIndex(of: "]") else { return false }
                out.append(MPiece(text: String(s[s.index(after: i)..<close]), script: 1))
                i = s.index(after: close)
            }
            var a: [MPiece] = []
            if !readArg(into: &a) { return false }
            out.append(MPiece(text: "√("))
            out.append(contentsOf: a)
            out.append(MPiece(text: ")"))
        case "text", "mathrm", "operatorname", "mathbf":
            guard i < s.endIndex, s[i] == "{", let close = s[i...].firstIndex(of: "}") else { return false }
            out.append(MPiece(text: String(s[s.index(after: i)..<close])))
            i = s.index(after: close)
        default:
            if let symbol = greekMap[name] {
                out.append(MPiece(text: symbol))
                return true
            }
            if let symbol = operatorMap[name] {
                out.append(MPiece(text: symbol))
                return true
            }
            return false // unknown command → caller degrades to literal source
        }
        return true
    }
}

/// Renders latex-lite to a styled Text, or nil when any part is unsupported —
/// the caller then renders the ORIGINAL expression string verbatim.
func renderMathText(_ latex: String, baseSize: CGFloat, color: Color) -> Text? {
    guard var attributed = renderMathAttributedString(latex, baseSize: baseSize, color: color) else { return nil }
    attributed.font = .system(size: baseSize)
    return Text(attributed)
}

/// AttributedString form — used for INLINE math inside a paragraph (a Text
/// cannot be embedded in an AttributedString). Baseline shifts ride the
/// SwiftUI `baselineOffset` attribute; sub/superscript runs carry smaller fonts.
func renderMathAttributedString(_ latex: String, baseSize: CGFloat, color: Color) -> AttributedString? {
    let normalized = normalizeLatex(latex)
    if normalized.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return nil }
    var parser = MathParser(normalized)
    if !parser.parse(stopBrace: false) { return nil }
    if parser.out.isEmpty { return nil }

    var result = AttributedString()
    for piece in parser.out {
        if piece.text.isEmpty { continue }
        var run = AttributedString(piece.text)
        run.font = .system(size: piece.script == 0 ? baseSize : baseSize * 0.8)
        run.foregroundColor = color
        if piece.italic {
            run.inlinePresentationIntent = .emphasized
        }
        switch piece.script {
        case 1: run.swiftUI.baselineOffset = baseSize * 0.33
        case -1: run.swiftUI.baselineOffset = -baseSize * 0.22
        default: break
        }
        result.append(run)
    }
    return result
}

File: android/app/src/main/java/com/grapsee/gsai/ui/chat/content/MathText.kt
package com.grapsee.gsai.ui.chat.content

import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.style.BaselineShift
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.sp
import com.grapsee.gsai.ui.theme.GsColors

/**
 * Native latex-lite rendering — no WebView, no remote service, no dependencies.
 *
 * Supports the notation AI assistants emit most: \frac, \sqrt, ^{} _{} (and
 * single-token forms), the Greek alphabet and a wide operator set, mapped to
 * Unicode. Superscript/subscript ride SpanStyle baseline shifts; Latin letters
 * render italic; digits and operators stay upright.
 *
 * Graceful degradation is the contract: any UNKNOWN \command makes
 * [renderMath] return null and the caller shows the ORIGINAL expression
 * verbatim — a wrong-looking formula is worse than an honest literal one.
 */

private data class MPiece(val text: String, val script: Int = 0, val italic: Boolean = false)

private val greekMap = mapOf(
    "alpha" to "α", "beta" to "β", "gamma" to "γ", "delta" to "δ", "epsilon" to "ε",
    "varepsilon" to "ε", "zeta" to "ζ", "eta" to "η", "theta" to "θ", "vartheta" to "ϑ",
    "iota" to "ι", "kappa" to "κ", "lambda" to "λ", "mu" to "μ", "nu" to "ν", "xi" to "ξ",
    "omicron" to "ο", "pi" to "π", "varpi" to "ϖ", "rho" to "ρ", "varrho" to "ϱ",
    "sigma" to "σ", "varsigma" to "ς", "tau" to "τ", "upsilon" to "υ", "phi" to "φ",
    "varphi" to "ϕ", "chi" to "χ", "psi" to "ψ", "omega" to "ω",
    "Gamma" to "Γ", "Delta" to "Δ", "Theta" to "Θ", "Lambda" to "Λ", "Xi" to "Ξ",
    "Pi" to "Π", "Sigma" to "Σ", "Upsilon" to "Υ", "Phi" to "Φ", "Psi" to "Ψ", "Omega" to "Ω"
)

private val operatorMap = mapOf(
    "times" to "×", "cdot" to "·", "div" to "÷", "pm" to "±", "mp" to "∓",
    "leq" to "≤", "geq" to "≥", "neq" to "≠", "ne" to "≠", "approx" to "≈",
    "equiv" to "≡", "sim" to "∼", "simeq" to "≃", "propto" to "∝", "infty" to "∞",
    "sum" to "∑", "prod" to "∏", "int" to "∫", "iint" to "∬", "oint" to "∮",
    "partial" to "∂", "nabla" to "∇", "in" to "∈", "notin" to "∉", "subset" to "⊂",
    "subseteq" to "⊆", "supset" to "⊃", "supseteq" to "⊇", "cup" to "∪", "cap" to "∩",
    "forall" to "∀", "exists" to "∃", "nexists" to "∤", "emptyset" to "∅",
    "varnothing" to "∅", "rightarrow" to "→", "to" to "→", "leftarrow" to "←",
    "gets" to "←", "leftrightarrow" to "↔", "Rightarrow" to "⇒", "Leftarrow" to "⇐",
    "Leftrightarrow" to "⇔", "mapsto" to "↦", "uparrow" to "↑", "downarrow" to "↓",
    "circ" to "∘", "bullet" to "•", "star" to "⋆", "ast" to "∗", "angle" to "∠",
    "degree" to "°", "perp" to "⊥", "parallel" to "∥", "land" to "∧", "lor" to "∨",
    "neg" to "¬", "hbar" to "ℏ", "ell" to "ℓ", "aleph" to "ℵ", "wp" to "℘",
    "Re" to "ℜ", "Im" to "ℑ", "prime" to "′", "langle" to "⟨", "rangle" to "⟩",
    "lceil" to "⌈", "rceil" to "⌉", "lfloor" to "⌊", "rfloor" to "⌋",
    "cdots" to "⋯", "ldots" to "…", "vdots" to "⋮", "ddots" to "⋱", "over" to "/"
)

/** Pre-normalization: drop spacing/limits commands and delimiters wrappers. */
private fun normalizeLatex(src: String): String = src
    .replace("\\left", "")
    .replace("\\right", "")
    .replace("\\displaystyle", "")
    .replace("\\limits", "")
    .replace("\\,", " ")
    .replace("\\;", " ")
    .replace("\\:", " ")
    .replace("\\!", "")

private class MathParser(private val s: String) {
    var i = 0
    val out = mutableListOf<MPiece>()

    /** Parses until end (or the matching `}` when [stopBrace]); false = unsupported. */
    fun parse(stopBrace: Boolean): Boolean {
        while (i < s.length) {
            val c = s[i]
            when {
                c == '}' -> {
                    if (stopBrace) return true
                    i++
                }
                c == '{' -> {
                    i++
                    if (!parse(true)) return false
                }
                c == '^' || c == '_' -> {
                    i++
                    val script = if (c == '^') 1 else -1
                    val arg = mutableListOf<MPiece>()
                    if (!readArg(arg)) return false
                    out += arg.map { MPiece(it.text, script, it.italic) }
                }
                c == '\\' -> {
                    if (!parseCommand()) return false
                }
                c.isLetter() -> {
                    out += MPiece(c.toString(), italic = true)
                    i++
                }
                c.isWhitespace() -> { out += MPiece(" "); i++ }
                else -> { out += MPiece(c.toString()); i++ }
            }
        }
        return true
    }

    /** Reads a {group} or single-token argument into [sink]. */
    private fun readArg(sink: MutableList<MPiece>): Boolean {
        if (i >= s.length) return false
        if (s[i] == '{') {
            i++
            val saved = out.size
            if (!parse(true)) return false
            sink += out.drop(saved)
            return true
        }
        if (s[i] == '\\') {
            val saved = out.size
            if (!parseCommand()) return false
            sink += out.drop(saved)
            return true
        }
        sink += MPiece(s[i].toString(), italic = s[i].isLetter())
        i++
        return true
    }

    private fun parseCommand(): Boolean {
        if (i + 1 >= s.length) return false
        val start = i + 1
        var end = start
        while (end < s.length && s[end].isLetter()) end++
        val name = s.substring(start, end)
        i = end
        when (name) {
            "frac" -> {
                val a = mutableListOf<MPiece>()
                val b = mutableListOf<MPiece>()
                if (!readArg(a) || !readArg(b)) return false
                out += MPiece("(", italic = false)
                out += a
                out += MPiece(")/(", italic = false)
                out += b
                out += MPiece(")", italic = false)
            }
            "sqrt" -> {
                // Optional [n] root index → superscript before the radical.
                if (i < s.length && s[i] == '[') {
                    var close = s.indexOf(']', i)
                    if (close < 0) return false
                    out += MPiece(s.substring(i + 1, close), script = 1)
                    i = close + 1
                }
                val a = mutableListOf<MPiece>()
                if (!readArg(a)) return false
                out += MPiece("√(", italic = false)
                out += a
                out += MPiece(")", italic = false)
            }
            "text", "mathrm", "operatorname", "mathbf" -> {
                if (i >= s.length || s[i] != '{') return false
                var close = s.indexOf('}', i)
                if (close < 0) return false
                out += MPiece(s.substring(i + 1, close))
                i = close + 1
            }
            else -> {
                greekMap[name]?.let { out += MPiece(it); return true }
                operatorMap[name]?.let { out += MPiece(it); return true }
                return false // unknown command → caller degrades to literal source
            }
        }
        return true
    }
}

/**
 * Renders latex-lite to styled spans, or null when any part is unsupported —
 * the caller then renders the ORIGINAL expression string verbatim.
 */
fun renderMath(latex: String, baseFontSize: TextUnit, color: Color): AnnotatedString? {
    val normalized = normalizeLatex(latex)
    if (normalized.isBlank()) return null
    val parser = MathParser(normalized)
    if (!parser.parse(stopBrace = false)) return null
    if (parser.out.isEmpty()) return null
    return buildAnnotatedString {
        for (piece in parser.out) {
            if (piece.text.isEmpty()) continue
            val style = SpanStyle(
                color = color,
                fontSize = if (piece.script == 0) baseFontSize else (baseFontSize.value * 0.8f).sp,
                baselineShift = when (piece.script) {
                    1 -> BaselineShift(0.35f)
                    -1 -> BaselineShift(-0.25f)
                    else -> BaselineShift.None
                },
                fontStyle = if (piece.italic) FontStyle.Italic else FontStyle.Normal
            )
            withStyle(style) { append(piece.text) }
        }
    }
}
